### PR TITLE
Import the new inline Data implementation and DataProtocol from the overlay

### PR DESCRIPTION
**Pull request by phausler**
This brings Data in sync with the overlay changes for the new inline revamp.

### CODE REVIEW

**Comment by phausler** [comment]
@swift-ci please test

File: Foundation/DataProtocol.swift [hunk 1]
```diff
@@ -0,0 +1,299 @@
...
+
+//===--- DataProtocol Extensions ------------------------------------------===//
+
+extension DataProtocol {
+    public func firstRange<D: DataProtocol>(of data: D) -> Range<Index>? {
+        return self.firstRange(of: data, in: self.startIndex ..< self.endIndex)
+    }
+
+    public func lastRange<D: DataProtocol>(of data: D) -> Range<Index>? {
+        return self.firstRange(of: data, in: self.startIndex ..< self.endIndex)
```
Comment (karwa) [inline Foundation/DataProtocol.swift]: Should be `lastRange`. I think this was fixed in the 5.0 cherry-pick but not on master.

File: Foundation/Data.swift [hunk 2]
```diff
@@ -791,229 +521,1349 @@ public final class _DataStorage {
...
+        @inlinable
+        mutating func resetBytes(in range: Range<Index>) {
+            assert(range.lowerBound <= MemoryLayout<Buffer>.size)
+            assert(range.upperBound <= MemoryLayout<Buffer>.size)
+            precondition(range.lowerBound <= length, "index \(range.lowerBound) is out of bounds of 0..<\(length)")
+            if count < range.upperBound {
+                count = range.upperBound
+            }
+            Swift.withUnsafeMutableBytes(of: &bytes) { rawBuffer in
+                bzero(rawBuffer.baseAddress?.advanced(by: range.lowerBound), range.upperBound - range.lowerBound)
```
Comment (karwa) [inline Foundation/Data.swift]: Similarly, this was replaced with memset. bzero isn't portable.

File: Foundation/DataProtocol.swift (hunk 1 above)
Comment (phausler) [reply Foundation/DataProtocol.swift]: yea we definitely shouldn't have master be different for Data and swift-5.0

**Comment by millenomi** [comment]
@swift-ci please test

**Comment by phausler** [comment]
This is out of date and supplanted by the other merge I landed.

**Comment by phausler** [comment]
#1834 was the one that landed

**Pull request closed without merging**